File: frontend/src/app/core/services/master-data.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, map } from 'rxjs';
import { environment } from '../../../environments/environment';
import {
  UOM, RawMaterialCategory,
  RawMaterial, FinishedGood, BOM, Customer, Supplier,
} from '../models/master-data.model';

@Injectable({ providedIn: 'root' })
export class MasterDataService {
  private http = inject(HttpClient);
  private base = `${environment.apiUrl}/master-data`;

  // Lookups
  listUOM(): Observable<UOM[]> {
    return this.http.get<{ data: UOM[] }>(`${this.base}/uom`).pipe(map(r => r.data));
  }

  listCategories(): Observable<RawMaterialCategory[]> {
    return this.http.get<{ data: RawMaterialCategory[] }>(`${this.base}/raw-material-categories`).pipe(map(r => r.data));
  }

  // Raw Materials
  listRawMaterials(): Observable<RawMaterial[]> {
    return this.http.get<{ data: RawMaterial[] }>(`${this.base}/raw-materials`).pipe(map(r => r.data));
  }

  getRawMaterial(id: string): Observable<RawMaterial> {
    return this.http.get<RawMaterial>(`${this.base}/raw-materials/${id}`);
  }

  createRawMaterial(body: object): Observable<RawMaterial> {
    return this.http.post<RawMaterial>(`${this.base}/raw-materials`, body);
  }

  updateRawMaterial(id: string, body: object): Observable<RawMaterial> {
    return this.http.put<RawMaterial>(`${this.base}/raw-materials/${id}`, body);
  }

  setRawMaterialActive(id: string, isActive: boolean): Observable<void> {
    return this.http.put<void>(`${this.base}/raw-materials/${id}/active`, { is_active: isActive });
  }

  // Finished Goods
  listFinishedGoods(): Observable<FinishedGood[]> {
    return this.http.get<{ data: FinishedGood[] }>(`${this.base}/finished-goods`).pipe(map(r => r.data));
  }

  getFinishedGood(id: string): Observable<FinishedGood> {
    return this.http.get<FinishedGood>(`${this.base}/finished-goods/${id}`);
  }

  createFinishedGood(body: object): Observable<FinishedGood> {
    return this.http.post<FinishedGood>(`${this.base}/finished-goods`, body);
  }

  updateFinishedGood(id: string, body: object): Observable<FinishedGood> {
    return this.http.put<FinishedGood>(`${this.base}/finished-goods/${id}`, body);
  }

  setFinishedGoodActive(id: string, isActive: boolean): Observable<void> {
    return this.http.put<void>(`${this.base}/finished-goods/${id}/active`, { is_active: isActive });
  }

  // BOM
  listBOMs(): Observable<BOM[]> {
    return this.http.get<{ data: BOM[] }>(`${this.base}/bom`).pipe(map(r => r.data));
  }

  getBOM(id: string): Observable<BOM> {
    return this.http.get<BOM>(`${this.base}/bom/${id}`);
  }

  createBOM(body: object): Observable<BOM> {
    return this.http.post<BOM>(`${this.base}/bom`, body);
  }

  updateBOM(id: string, body: object): Observable<BOM> {
    return this.http.put<BOM>(`${this.base}/bom/${id}`, body);
  }

  // Customers
  listCustomers(): Observable<Customer[]> {
    return this.http.get<{ data: Customer[] }>(`${this.base}/customers`).pipe(map(r => r.data));
  }

  getCustomer(id: string): Observable<Customer> {
    return this.http.get<Customer>(`${this.base}/customers/${id}`);
  }

  createCustomer(body: object): Observable<Customer> {
    return this.http.post<Customer>(`${this.base}/customers`, body);
  }

  updateCustomer(id: string, body: object): Observable<Customer> {
    return this.http.put<Customer>(`${this.base}/customers/${id}`, body);
  }

  setCustomerActive(id: string, isActive: boolean): Observable<void> {
    return this.http.put<void>(`${this.base}/customers/${id}/active`, { is_active: isActive });
  }

  // Suppliers
  listSuppliers(): Observable<Supplier[]> {
    return this.http.get<{ data: Supplier[] }>(`${this.base}/suppliers`).pipe(map(r => r.data));
  }

  getSupplier(id: string): Observable<Supplier> {
    return this.http.get<Supplier>(`${this.base}/suppliers/${id}`);
  }

  createSupplier(body: object): Observable<Supplier> {
    return this.http.post<Supplier>(`${this.base}/suppliers`, body);
  }

  updateSupplier(id: string, body: object): Observable<Supplier> {
    return this.http.put<Supplier>(`${this.base}/suppliers/${id}`, body);
  }

  setSupplierActive(id: string, isActive: boolean): Observable<void> {
    return this.http.put<void>(`${this.base}/suppliers/${id}/active`, { is_active: isActive });
  }
}

File: backend/internal/service/master_data_service.go
package service

import (
	"context"

	"github.com/water-factory/api/internal/domain"
	"github.com/water-factory/api/internal/dto"
	"github.com/water-factory/api/internal/repository"
)

type MasterDataService struct {
	repo *repository.MasterDataRepo
}

func NewMasterDataService(repo *repository.MasterDataRepo) *MasterDataService {
	return &MasterDataService{repo: repo}
}

func (s *MasterDataService) ListUOM(ctx context.Context) ([]*domain.UnitOfMeasure, error) {
	return s.repo.ListUOM(ctx)
}

func (s *MasterDataService) ListCategories(ctx context.Context) ([]*domain.RawMaterialCategory, error) {
	return s.repo.ListCategories(ctx)
}

// Raw Materials

func (s *MasterDataService) ListRawMaterials(ctx context.Context) ([]*domain.RawMaterial, error) {
	return s.repo.ListRawMaterials(ctx)
}

func (s *MasterDataService) GetRawMaterial(ctx context.Context, id string) (*domain.RawMaterial, error) {
	return s.repo.GetRawMaterial(ctx, id)
}

func (s *MasterDataService) CreateRawMaterial(ctx context.Context, req *dto.CreateRawMaterialRequest) (*domain.RawMaterial, error) {
	return s.repo.CreateRawMaterial(ctx, req)
}

func (s *MasterDataService) UpdateRawMaterial(ctx context.Context, id string, req *dto.UpdateRawMaterialRequest) (*domain.RawMaterial, error) {
	return s.repo.UpdateRawMaterial(ctx, id, req)
}

func (s *MasterDataService) SetRawMaterialActive(ctx context.Context, id string, active bool) error {
	return s.repo.SetRawMaterialActive(ctx, id, active)
}

// Finished Goods

func (s *MasterDataService) ListFinishedGoods(ctx context.Context) ([]*domain.FinishedGood, error) {
	return s.repo.ListFinishedGoods(ctx)
}

func (s *MasterDataService) GetFinishedGood(ctx context.Context, id string) (*domain.FinishedGood, error) {
	return s.repo.GetFinishedGood(ctx, id)
}

func (s *MasterDataService) CreateFinishedGood(ctx context.Context, req *dto.CreateFinishedGoodRequest) (*domain.FinishedGood, error) {
	return s.repo.CreateFinishedGood(ctx, req)
}

func (s *MasterDataService) UpdateFinishedGood(ctx context.Context, id string, req *dto.UpdateFinishedGoodRequest) (*domain.FinishedGood, error) {
	return s.repo.UpdateFinishedGood(ctx, id, req)
}

func (s *MasterDataService) SetFinishedGoodActive(ctx context.Context, id string, active bool) error {
	return s.repo.SetFinishedGoodActive(ctx, id, active)
}

// BOM

func (s *MasterDataService) ListBOMs(ctx context.Context) ([]*domain.BOM, error) {
	return s.repo.ListBOMs(ctx)
}

func (s *MasterDataService) GetBOM(ctx context.Context, id string) (*domain.BOM, error) {
	return s.repo.GetBOM(ctx, id)
}

func (s *MasterDataService) CreateBOM(ctx context.Context, req *dto.CreateBOMRequest) (*domain.BOM, error) {
	return s.repo.CreateBOM(ctx, req)
}

func (s *MasterDataService) UpdateBOM(ctx context.Context, id string, req *dto.UpdateBOMRequest) (*domain.BOM, error) {
	return s.repo.UpdateBOM(ctx, id, req)
}

// Customers

func (s *MasterDataService) ListCustomers(ctx context.Context) ([]*domain.Customer, error) {
	return s.repo.ListCustomers(ctx)
}

func (s *MasterDataService) GetCustomer(ctx context.Context, id string) (*domain.Customer, error) {
	return s.repo.GetCustomer(ctx, id)
}

func (s *MasterDataService) CreateCustomer(ctx context.Context, req *dto.CreateCustomerRequest) (*domain.Customer, error) {
	return s.repo.CreateCustomer(ctx, req)
}

func (s *MasterDataService) UpdateCustomer(ctx context.Context, id string, req *dto.UpdateCustomerRequest) (*domain.Customer, error) {
	return s.repo.UpdateCustomer(ctx, id, req)
}

func (s *MasterDataService) SetCustomerActive(ctx context.Context, id string, active bool) error {
	return s.repo.SetCustomerActive(ctx, id, active)
}

// Suppliers

func (s *MasterDataService) ListSuppliers(ctx context.Context) ([]*domain.Supplier, error) {
	return s.repo.ListSuppliers(ctx)
}

func (s *MasterDataService) GetSupplier(ctx context.Context, id string) (*domain.Supplier, error) {
	return s.repo.GetSupplier(ctx, id)
}

func (s *MasterDataService) CreateSupplier(ctx context.Context, req *dto.CreateSupplierRequest) (*domain.Supplier, error) {
	return s.repo.CreateSupplier(ctx, req)
}

func (s *MasterDataService) UpdateSupplier(ctx context.Context, id string, req *dto.UpdateSupplierRequest) (*domain.Supplier, error) {
	return s.repo.UpdateSupplier(ctx, id, req)
}

func (s *MasterDataService) SetSupplierActive(ctx context.Context, id string, active bool) error {
	return s.repo.SetSupplierActive(ctx, id, active)
}
